src/repo/fs.rs

```diff
@@ -5,147 +5,462 @@ use async_trait::async_trait;
+                file.read_to_end(&mut data)?;
+                let block = Block::new(data.into_boxed_slice(), cid);
+                Ok(Some(block))
+            })
+            .await?
+        }
+        .instrument(span)
+        .await
     }
 
     async fn put(&self, block: Block) -> Result<(Cid, BlockPut), Error> {
-        let path = block_path(self.path.clone(), &block.cid());
-        let cids = &self.cids;
-        let data = block.data();
-        let mut file = fs::File::create(path).await?;
-        file.write_all(&*data).await?;
-        file.flush().await?;
-        let retval = if cids.lock().await.insert(RepoCid(block.cid().to_owned())) {
-            BlockPut::NewBlock
-        } else {
-            BlockPut::Existed
-        };
-        // FIXME: checking if the file existed already while creating complicates this function a
-        // lot; might be better to just guard with mutex to enforce single task file access.. the
-        // current implementation will write over the same file multiple times, each time believing
-        // it was the first.
-        Ok((block.cid().to_owned(), retval))
+        use std::collections::hash_map::Entry;
+
+        let span = tracing::trace_span!("put block", cid = %block.cid());
+
+        let target_path = block_path(self.path.clone(), &block.cid());
+        let cid = block.cid;
+        let data = block.data;
+
+        let inner_span = span.clone();
+
+        async move {
+            // why synchronize here? because when we lose the race we cant know if there was someone
+            // else interested in writing this block or not
+
+            // FIXME: allowing only the creator to cleanup means this is not forget safe. the forget
+            // doesn't result in memory unsafety but it will deadlock any other access to the cid.
+            let (tx, mut rx) = {
+                let mut g = self.writes.lock().expect("cant support poisoned");
+
+                match g.entry(RepoCid(cid.to_owned())) {
+                    Entry::Occupied(oe) => {
+                        // someone is already writing this, nice
+                        trace!("joining in on another already writing the block");
+                        (oe.get().clone(), oe.get().subscribe())
+                    }
+                    Entry::Vacant(ve) => {
+                        // we might be the first, or then the block exists already
+                        let (tx, rx) = broadcast::channel(1);
+                        ve.insert(tx.clone());
+                        (tx, rx)
+                    }
+                }
+            };
+
+            // create this in case the winner is dropped while awaiting
+            let cleanup = RemoveOnDrop(self.writes.clone(), Some(RepoCid(cid.to_owned())));
+
+            // launch a blocking task for the filesystem mutation.
+            // `tx` is moved into the task but `rx` stays in the async context.
+            let je = tokio::task::spawn_blocking(move || {
```

Comment:
What is `je`?

Reply:
this was `Result<_, JoinError>` at one point, the most outer result. Could be `jh` as well for JoinHandle :)

Reply:
yeah totally a nit but why not just `join_handle`? Maybe there's some rustlang convention I don't know about :shrug: